### ISSUE TITLE
Source-build repo legs should support previously source-built artifacts

### ISSUE THREAD

**Issue by mmitche** [dotnet/source-build#3479]
### Describe the Problem

Right now, repo source build legs have 3 options for input dependencies:
- Satisfying them as with ref-only SBRPs
- Satisfying them with source build intermediates via flow from other repos
- Satisfying them via existing product/public feeds, and excluding them from prebuilt reports

The last category must be used for dependencies that will be satisfied in full source-build by previously source-built artifacts. Primarily, build-time dependencies of repos low in the stack (e.g. an arcade dependency on a runtime OOB package).

This is a non-ideal way of dealing with these kinds of dependencies. It makes choosing an input version difficult. Repo builds build may wish to maintain an older version of the input dependency, perhaps for developer workflow reasons (e.g. not getting ahead of a version used in VS). Unfortunately, staying on an older version does open up the possibility that when built in full source-build, there will be breaks. So a repo might be left with the only good option being to add custom logic and choose one version in repo source build and another in Microsoft's build. Even with such logic, it falls on the repo owner to keep the dependency up to date, and this potentially introduces additional unwanted flow into the product (e.g. arcade should not take regular updates from runtime).

### Describe the Solution

I think the right way to solve this is to introduce previously source-built artifacts as an input to repo source-build. These would be updated and flowed via Maestro. Additionally, the previously source built artifacts would override the versions of input dependencies to a repo, though only for those dependencies built after or concurrently in the source-build graph. Dependencies that come in via previously source built must be classified as Toolset dependencies in `eng/Version.Details.xml`

As an example, if arcade has a dependency on a runtime asset, and cannot use an SBRP. it would list that dependency in `eng/Version.Details.xml`, under toolset. It may choose even an old stable version. Additionally, a dependency for previously source-built artifacts would appear. During the build, the PSBA would be unpacked, and an input Versions.props would be constructed. It may be simplest to only include those version numbers for dependencies that appear in the toolset dependencies of a repo, or perhaps the logic is smart enough to filter based on what builds before the repo in the SB graph. Those versions would override the default repo version.

**Comment by mthalman** [comment]
> Unfortunately, staying on an older version does open up the possibility that when built in full source-build, there will be breaks. So a repo might be left with the only good option being to add custom logic and choose one version in repo source build and another in Microsoft's build.

Once per-repo-PVP is fully implemented, there's really no difference here between source-build and MS build, right? They'd both compile against the version declared in the repo.

**Comment by mmitche** [comment]
Full source-build would still pull those from previously source-built in cases where the input dep is built after the current repo.

**Comment by MichaelSimons** [comment]
Is https://github.com/dotnet/arcade/pull/13699 a concrete use case for this?  I ask because in that situation arcade needs the latest released stable version which is different than source-build N-1 during a preview cycle.

**Comment by mmitche** [comment]
> Is [dotnet/arcade#13699](https://github.com/dotnet/arcade/pull/13699) a concrete use case for this? I ask because in that situation arcade needs the latest released stable version which is different than source-build N-1 during a preview cycle.

Yes. We don't want to use N-1 in a developer context, but we should validate against it in source-build.

**Comment by MichaelSimons** [comment]
Closing as repo level prebuilt detection is being removed with https://github.com/dotnet/source-build/issues/5034.  There is no longer value in doing this.
